### PR TITLE
[Merged by Bors] - feat: port AlgebraicTopology.SimplicialObject

### PR DESCRIPTION
**Pull request by joelriou**
---
- [x] depends on: #3286
- [x] depends on: #3340 

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#3286
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#3286~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

File: Mathlib/AlgebraicTopology/SimplicialObject.lean [hunk 1]
```diff
@@ -0,0 +1,874 @@
...
+variable (C : Type u) [Category.{v} C]
+
+-- porting note: removed @[nolint has_nonempty_instance]
+/-- The category of simplicial objects valued in a category `C`.
+This is the category of contravariant functors from `SimplexCategory` to `C`. -/
+def SimplicialObject :=
+  SimplexCategoryᵒᵖ ⥤ C
+#align category_theory.simplicial_object CategoryTheory.SimplicialObject
+
+@[simps!]
```
Comment (ChrisHughes24) [inline Mathlib/AlgebraicTopology/SimplicialObject.lean:47]: I'm a bit confused about all these `simps!` attributes. There was nothing similar in mathlib3 right? Are they strictly necessary, and would it perhaps be better to make `SimplicialObject` a `reducible` instead?

Comment (joelriou) [reply Mathlib/AlgebraicTopology/SimplicialObject.lean]: Without the `reducible` attribute, the `simps!` for the Category instance (and extra `ext` lemmas for morphisms) were necessary. I have made `SimplicialObject` reducible in the last commit.

Comment (ChrisHughes24) [reply Mathlib/AlgebraicTopology/SimplicialObject.lean]: Okay, I'll ping @jcommelin , @semorrison and @adamtopaz to see if they're happy with the change.

Comment (semorrison) [reply Mathlib/AlgebraicTopology/SimplicialObject.lean]: I mean, I'd prefer that we don't make it reducible. It's more work to set up here, unfortunately. We have a tendency to regret making things less opaque, however...

Comment (joelriou) [reply Mathlib/AlgebraicTopology/SimplicialObject.lean:47]: Ok. I have reverted back to the previous situation where `SimplicialObject` is not declared reducible.

Comment (ChrisHughes24) [reply Mathlib/AlgebraicTopology/SimplicialObject.lean:47]: I'll let someone else take over the review of this PR, but I'm not sure that `simp` lemmas unfolding the definition of a type is a particularly good idea either for the same reasons. Probably a lot of the simp lemmas that become accessible after unfolding definitions should be proved on `SimplicialObject` directly.

Comment (joelriou) [reply Mathlib/AlgebraicTopology/SimplicialObject.lean:47]: Indeed, a lemma like
```lean
theorem SimplicialObject.δ_naturality {X' X : SimplicialObject C} (f : X ⟶ X') {n : ℕ} (i : Fin (n + 2)) :
    X.δ i ≫ f.app (op [n]) = f.app (op [n + 1]) ≫ X'.δ i```
is a particular case of `NatTrans.naturality`, and it is stated as a `simp` lemma that is specific to `SimplicialObject`.

Comment (semorrison) [reply Mathlib/AlgebraicTopology/SimplicialObject.lean:47]: Sorry, a bit of a tangle here. I have a PR open on mathlib proposing to adjust the simps projections for `category` so we don't unfold `Hom`. I should have got that done and forwarded port faster...

Comment (semorrison) [reply Mathlib/AlgebraicTopology/SimplicialObject.lean:47]: Okay, I propose that I will quickly forward port that PR, and then merge that into this before we proceed?

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#3286~~
* ~~leanprover-community/mathlib4#3340~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by ChrisHughes24** [comment]
bors merge

**Pull request closed without merging**